File: app/model/competency.rb
# frozen_string_literal: true

require_relative 'base'
require_relative 'competency_level'

class Competency < Base
  attr_accessor :title, :description, :id, :filtered_levels

  def initialize(competency)
    @id = competency['id']
    @title = competency['title']
    @description = competency['description']
  end

  def competency_levels
    @competency_levels ||= CompetencyLevel.find_by(competency_id: @id)
  end

  def competency_levels_by_semester(semester)
    CompetencyLevel.find_by(competency_id: @id, semester: semester)
  end

  class << self
    def plural
      'competencies'
    end

    def find_by(topic_id:)
      all.select do |competency|
        topic_id == competency.id[0]
      end
    end
  end
end

File: src/main/java/model/Competency.java
package model;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.ArrayList;

@NoArgsConstructor
@Getter
@Setter
public class Competency {
    private String id;
    private String title;
    private String description;
    private ArrayList<CompetencyLevel> competencyLevels;

    public Competency(String id, String title, String description) {
        this.id = id;
        this.title = title;
        this.description = description;
        this.competencyLevels = new ArrayList<>();
    }

    public void addCompetencyLevel(CompetencyLevel level) {
        this.competencyLevels.add(level);
    }

}
